File: sandbox/demo/args.py
import argparse
from dataclasses import dataclass

from sandbox.demo.config import Config
from sandbox.demo.network import NetworkMode


@dataclass(frozen=True, slots=True)
class Args:
    config: Config
    verbose: bool


def parse_args(argv: list[str]) -> Args:
    p = argparse.ArgumentParser()

    # sandbox config args
    p.add_argument(
        "--network-mode",
        type=NetworkMode,
        default=NetworkMode.HOST,
        help='Set to "none" to disable internet access. Default: "host" (no network restrictions).',
    )
    p.add_argument(
        "command",
        nargs="+",
        help="Command to execute inside the sandbox.",
    )

    # sandbox tool args
    p.add_argument("--verbose", action="store_true")

    ns = p.parse_args(argv)

    config = Config(
        network_mode=ns.network_mode,
        command=ns.command,
    )

    return Args(
        config=config,
        verbose=ns.verbose,
    )

File: sandbox/demo/args.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"io"
	"strings"

	"sandbox/demo/config"
	"sandbox/demo/network"
)

type Args struct {
	Config  config.Config
	Verbose bool
}

func ParseArgs(argv []string) (Args, error) {
	fs := flag.NewFlagSet("sandbox", flag.ContinueOnError)
	fs.SetOutput(io.Discard)

	networkModeStr := fs.String("network-mode", string(network.Host), "")
	verbose := fs.Bool("verbose", false, "")

	if err := fs.Parse(argv); err != nil {
		return Args{}, err
	}

	command := fs.Args()
	if len(command) == 0 {
		return Args{}, errors.New("error: the following arguments are required: command")
	}

	for _, arg := range command {
		if strings.TrimSpace(arg) == "" {
			return Args{}, errors.New("value_error: command argument is empty or only whitespace")
		}
	}

	netMode := network.NetworkMode(*networkModeStr)
	if !netMode.IsValid() {
		return Args{}, fmt.Errorf("value_error: %q is not a valid NetworkMode", *networkModeStr)
	}

	cfg := config.Config{
		NetworkMode: netMode,
		Command:     command,
	}

	return Args{
		Config:  cfg,
		Verbose: *verbose,
	}, nil
}
